Bottom: ced064e361cd55883d55df3ebca12b1e89d750f0
Top:    bbea19cf7078756a8ec64798b79973d24550af66
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-10-21 19:45:06 +0100

Give callback from NoHolesPolygonDicer ownership of the returned contour

This means callers of NoHolesPolygonDicer() should call poly_FreeContours
on the contour they are passed (if they do not wish to retain it).


---

diff --git a/src/hid/common/draw_helpers.c b/src/hid/common/draw_helpers.c
index 1b7c6b7..a7c4620 100644
--- a/src/hid/common/draw_helpers.c
+++ b/src/hid/common/draw_helpers.c
@@ -57,7 +57,10 @@ static void thindraw_contour (hidGC gc, PLINE *pl)
 static void fill_contour_cb (PLINE *pl, void *user_data)
 {
   hidGC gc = user_data;
+  PLINE *local_pl = pl;
+
   fill_contour (gc, pl);
+  poly_FreeContours (&local_pl);
 }
 
 void common_fill_pcb_polygon (hidGC gc, PolygonType *poly,
diff --git a/src/polygon.c b/src/polygon.c
index d5df073..46cba47 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -1517,6 +1517,7 @@ r_NoHolesPolygonDicer (PLINE * p,
   pa->contours = p;
   if (!p->next)                 /* no holes */
     {
+      pa->contours = NULL; /* The callback now owns the contour */
       emit (p, user_data);
       poly_Free (&pa);
       return;
